### PR TITLE
BUG: fix a numpy.npiter leak in npyiter_multi_index_set

### LINKED ISSUE #19375: Memory leak in function `npyiter_multi_index_set` (a static analysis report with a POC)
### Reproducing code example:

```py
import sys

import numpy as np

index = 0
it = np.nditer(np.array([43124, 1234, 123432, 123432]), flags=['multi_index'])

print(sys.getrefcount(index))

for i in range(200):
    it.multi_index = (index,)

print(sys.getrefcount(index))
```
output:
```
1962
2162
```

And when replacing the array with `[10000000000] * (index + 1)` and `index = 10000`, the output will be changed to `4` and `204` for the two prints.

As the reference count linearly increases with the number of invocations, which indicates a memory leak.

### Error message:

The path provided by the static analyzer is as follows.

1. A new reference returned from [`PySequence_GetItem`](https://docs.python.org/3/c-api/sequence.html#c.PySequence_GetItem) and pointed to be `v`.
https://github.com/numpy/numpy/blob/04ab04d93d4d7a4d241fe0ceb725436a8b6c8c2e/numpy/core/src/multiarray/nditer_pywrap.c#L1596

2. Call [`PyLong_AsLong`](https://docs.python.org/3/c-api/long.html#c.PyLong_AsLong) with `v`, which does not steal reference.
https://github.com/numpy/numpy/blob/04ab04d93d4d7a4d241fe0ceb725436a8b6c8c2e/numpy/core/src/multiarray/nditer_pywrap.c#L1597

3. Assume no error happens, take the false branch.
https://github.com/numpy/numpy/blob/04ab04d93d4d7a4d241fe0ceb725436a8b6c8c2e/numpy/core/src/multiarray/nditer_pywrap.c#L1598

4. Assume loop continues, the object pointed to by `v` is leaked without decreasing reference count.
https://github.com/numpy/numpy/blob/04ab04d93d4d7a4d241fe0ceb725436a8b6c8c2e/numpy/core/src/multiarray/nditer_pywrap.c#L1595

### NumPy/Python version information:

Static analysis carried out on commit 04ab04d.

### PR DESCRIPTION
**Pull request by defoishugo**
Resolves #19375. 

Moves the *Py_DECREF* call one line before, so the reference count will be always decresed.

## Example

```
import sys

import numpy as np

index = 0
it = np.nditer(np.array([43124, 1234, 123432, 123432]), flags=['multi_index'])

print(sys.getrefcount(index))

for i in range(200):
    it.multi_index = (index,)

print(sys.getrefcount(index))
```

## Before this PR

```
1962
2162
```

## After this PR

```
1403
1403
```

Note: this is my first PR on this repository and my second one in my life, tell me if you have any feedback.

### CODE REVIEW

**Comment by seberg** [comment]
Thanks!  @defoishugo would you be able to add a test for this code?  I expect we do not have any code coverage at all here.

The test does not have to check for the reference count error itself (it can, but it is not important to me).  Just covering the code paths would be awesome.  (I occasionally use `pytest-leaks` which should have noticed this if we had test coverage.

**Comment by defoishugo** [comment]
> 
> 
> Thanks! @defoishugo would you be able to add a test for this code? I expect we do not have any code coverage at all here.
> 
> The test does not have to check for the reference count error itself (it can, but it is not important to me). Just covering the code paths would be awesome. (I occasionally use `pytest-leaks` which should have noticed this if we had test coverage.

Thanks for your feedback @seberg.

Looking into *numpy/core/tests/test_nditer.py*, I can find some tests  using *multi_index* behaviour. Don't know yet why they are not catching leaks when you try pytest-leaks. I will add the test you are asking inside this file.

However, I cannot execute the runtests.py, because of a DeprecationWarning exception, that stop the execution of the tests. Do you recommend any manipulation on my side? I installed the requirements again as in the testing guidelines and already tried to execute python with "-W ignore" parameter.

I will check sooner for this problem, but if you have any idea of what I can do it will be much quicker for me to add the appropriate test. If you don't have any idea, I will just get out of my anaconda environment to create another one from scratch.

**Comment by seberg** [comment]
> However, I cannot execute the runtests.py, because of a DeprecationWarning exception, that stop the execution of the tests

Hmmm, that is strange.  You could try `python3 runtests.py -- --continue-on-collection-errors`.  Maybe report that error, even if we just notice some incompatibility and close it again.

> Don't know yet why they are not catching leaks when you try pytest-leaks.

A quick look: I think we have tests, but it seems only for the error path, and not for the normal path.

**Comment by defoishugo** [comment]
> > However, I cannot execute the runtests.py, because of a DeprecationWarning exception, that stop the execution of the tests
> 
> Hmmm, that is strange. You could try `python3 runtests.py -- --continue-on-collection-errors`. Maybe report that error, even if we just notice some incompatibility and close it again.
> 
> > Don't know yet why they are not catching leaks when you try pytest-leaks.
> 
> A quick look: I think we have tests, but it seems only for the error path, and not for the normal path.

Thank you for your support @seberg.

The tests are implemented. Please tell me if anything should be edited. 

By the way, I will work as soon as possible on the Travis CI check, which is failing.

**Comment by seberg** [comment]
Just on mobile, but if its just one test failing, please ignore it. We have some flaky CI unfortunately.

**Comment by charris** [comment]
> We have some flaky CI unfortunately.

Yep, ARM64 is acting up again.

**Comment by defoishugo** [comment]
@mattip we could also merge this one I think

File: numpy/core/tests/test_nditer.py [hunk 1]
```diff
@@ -185,6 +185,29 @@ def test_iter_c_or_f_order():
...
+    it = np.nditer(a, flags=['multi_index'])
+
+    # Removes the iteration on two first elements of a[0]
+    it.multi_index = (0, 2,)
+
+    assert_equal([i for i in it], [2, 3, 4, 5])
+    
+@pytest.mark.skipif(not HAS_REFCOUNT, reason="Python lacks refcounts")
+def test_nditer_multi_index_set_refcount():
+    # Test if the reference count on inde variable is decreased
```
Comment (mattip) [inline numpy/core/tests/test_nditer.py:200]:
```suggestion
    # Test if the reference count on index variable is decreased
```

**Comment by mattip** [comment]
Thanks @defoishugo

**Pull request merged**